### PR TITLE
[SPARK-28519][SQL] Use StrictMath log, pow functions for platform independence

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

See discussion on the JIRA (and dev@). At heart, we find that math.log and math.pow can actually return slightly different results across platforms because of hardware optimizations. For the actual SQL log and pow functions, I propose that we should use StrictMath instead to ensure the answers are already the same. (This should have the benefit of helping tests pass on aarch64.)

Further, the atanh function (which is not part of java.lang.Math) can be implemented in a slightly different and more accurate way.

## How was this patch tested?

Existing tests (which will need to be changed).
Some manual testing locally to understand the numeric issues.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 1]
```diff
@@ -291,10 +291,6 @@ case class Cosh(child: Expression) extends UnaryMathExpression(math.cosh, "COSH"
   usage = """
     _FUNC_(expr) - Returns inverse hyperbolic cosine of `expr`.
   """,
-  arguments = """
-    Arguments:
-      * expr - hyperbolic angle
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala]: This description is wrong; the argument is not an angle. For consistency with other inverse functions, I just removed the description.

File: sql/core/src/test/resources/sql-tests/results/pgSQL/float8.sql.out [hunk 2]
```diff
@@ -528,7 +528,7 @@ SELECT atanh(double('0.5'))
 -- !query 56 schema
 struct<ATANH(CAST(0.5 AS DOUBLE)):double>
 -- !query 56 output
-0.5493061443340549
+0.5493061443340548
```
Comment (srowen) [inline sql/core/src/test/resources/sql-tests/results/pgSQL/float8.sql.out]: Other test values will need to change.
Takeshi says this isn't the value from PostgreSQL anyway

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala]: Oops. Thanks for the fix.

File: sql/core/src/test/resources/sql-tests/results/pgSQL/float8.sql.out (hunk 2 above)
Comment (maropu) [reply sql/core/src/test/resources/sql-tests/results/pgSQL/float8.sql.out]: Yeah, thanks! Looks nice.

**Comment by maropu** [comment]
I quickly checked performance numbers on my local;
```
// on Mac (2 GHz Intel Core i5)
scala> spark.range(30000000L).selectExpr("acosh(id)").write.format("noop").save()

w/Math
Elapsed time: 0.5992446991111111s

w/StrictMath
Elapsed time: 1.1835621365555555s
```
As you said in the mail, there are some overheads on x86/64.
Could you check them on power, @kiszk ?

**Comment by kiszk** [comment]
@maropu Thank you, I found [an entry](https://bugs.openjdk.java.net/browse/JDK-8210416) for StrictMath optimization in OpenJDK.   
Could you please add the version of OpenJDK that you used, too?

**Comment by maropu** [comment]
Updated above.

**Comment by kiszk** [comment]
Do we need to update `sql-migration-guide.md`?

File: sql/core/src/test/resources/sql-tests/results/pgSQL/float8.sql.out (hunk 2 above)
Comment (wangyum) [reply sql/core/src/test/resources/sql-tests/results/pgSQL/float8.sql.out]: Yes. PostgreSQL  returns `0.549306144334055`:
https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/expected/float8.out#L490-L494

**Comment by srowen** [comment]
See https://issues.apache.org/jira/browse/SPARK-28519?focusedCommentId=16895279&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16895279 for an interesting bit of analysis. The StrictMath answer does appear correct and the difference comes from some alternate assembly code implementation in the x86 JVM. Whatever happens there we may wish to sidestep the problem with a change like this anyway.

Yes I can update the migration guide with a note that the returned value may be very slightly different.

The perf overhead is trivial here, so that's no issue.

Now let me investigate the further test failures.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 3]
```diff
@@ -486,7 +483,7 @@ case class Factorial(child: Expression) extends UnaryExpression with ImplicitCas
       > SELECT _FUNC_(1);
        0.0
   """)
-case class Log(child: Expression) extends UnaryLogExpression(math.log, "LOG")
+case class Log(child: Expression) extends UnaryLogExpression(StrictMath.log, "LOG")
```
Comment (huangtianhua) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala]: Maybe we should modify math.exp to StrictMath.exp in line 361:
# on aarch64
scala> StrictMath.exp(1.0)
res4: Double = 2.7182818284590455

scala> math.exp(1.0)
res5: Double = 2.7182818284590455

# on x86_64:

scala> math.exp(1.0)
res44: Double = 2.718281828459045

scala> StrictMath.exp(1.0)
res45: Double = 2.7182818284590455

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala]: OK will do.

**Comment by huangtianhua** [comment]
@srowen Thanks a lot. And a good news:) all tests passed on x86_64 and aarch64 based on your pr(only two tests now we increase the timeout, later we will tests on larger server), see: https://logs.openlabtesting.org/logs/6/6/947ddad683ad7a2e0a0cc4c2310e352ace21a86f/check/spark-build-arm64/8e39061/

**Review by srowen — COMMENTED** [review]
OK I feel pretty good about the change and the reasoning. I'll leave it open a little longer for comments.

**Comment by srowen** [comment]
Merged to master

**Comment by gatorsmile** [comment]
@srowen Could you show the perf benchmark? The performance regression is expected, right?

**Comment by srowen** [comment]
I did not benchmark this as I think it's a correctness issue that would be worth a perf hit. I also expect it makes almost no difference - computing a function in SQL is dominated by so much more than the math here. Let me assess that though with some microbenchmarks.

**Comment by srowen** [comment]
@gatorsmile here's a quick benchmark for `log()`:

Previous:
```
Running benchmark: log
  Running case: log
  Stopped after 20 iterations, 6376 ms

Java HotSpot(TM) 64-Bit Server VM 1.8.0_192-b12 on Mac OS X 10.14.6
Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
log:                                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
log                                                 297            319          17          6.7         148.3       1.0X
```
Current:
```
Running benchmark: log
  Running case: log
  Stopped after 20 iterations, 6507 ms

Java HotSpot(TM) 64-Bit Server VM 1.8.0_192-b12 on Mac OS X 10.14.6
Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
log:                                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
log                                                 301            325          15          6.7         150.3       1.0X
```

Looks like it's slower by about 2 nanoseconds, but it's not even statistically significant.
Even StrictMath is native code. I think we could say there is no perf hit.

**Comment by gatorsmile** [comment]
Thanks! @srowen 

cc @rednaxelafx

**Comment by rednaxelafx** [comment]
Thanks @srowen ! Basically we just want to make sure if we anticipate there are performance implications, some due diligence of benchmarking should be done. Thanks for adding that information in this PR for future reference!

BTW, folks that are not intimately familiar with the inner workings of the HotSpot JVM may find this interesting:
While both `java.lang.Math` and `java.lang.StrictMath` are partially implemented with `native` functions, they're really different under the hood.

In current OpenJDK8u,
- java.lang.StrictMath: 
  - Java side: http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/89c8bfe10659/src/share/classes/java/lang/StrictMath.java
  - native side: http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/89c8bfe10659/src/share/native/java/lang/StrictMath.c
- java.lang.Math:
  - Java side: http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/89c8bfe10659/src/share/classes/java/lang/Math.java

Now, taking one specific example of `Math.log()`:
```java
    public static double log(double a) {
        return StrictMath.log(a); // default impl. delegates to StrictMath
    }
```
goes to `StrictMath.log()`:
```java
    public static native double log(double a);
```
goes to native code:
```c
JNIEXPORT jdouble JNICALL
Java_java_lang_StrictMath_log(JNIEnv *env, jclass unused, jdouble d)
{
    return (jdouble) jlog((double)d);
}
```
and then `jlog` is actually the `log` function from fdlibm via:
```c
src/share/native/java/lang/fdlibm/include/jfdlibm.h:44:#define log     jlog
```

So then `Math.log` and `StrictMath.log` should be the exact same thing? Well not so fast.
In the HotSpot JVM, it treats some Java methods as being "special", aka "intrinsics", which means it knows exactly what semantics that method needs to implement, and choose an alternative internal implementation for it.
http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/c7a3e57fdf4a/src/share/vm/classfile/vmSymbols.hpp#l681
```c++
  do_intrinsic(_dlog,                     java_lang_Math,         log_name,   double_double_signature,           F_S)   \
```
`Math.log()` happens to be one of the intrinsified Java methods. It is implemented with special assembler snippets in all of:
1. Interpreter, e.g.: http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/c7a3e57fdf4a/src/cpu/x86/vm/interpreter_x86_64.cpp#l269 -> http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/c7a3e57fdf4a/src/cpu/x86/vm/assembler_x86.cpp#l4147
2. C1 compiler, e.g.: http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/c7a3e57fdf4a/src/cpu/x86/vm/c1_LIRAssembler_x86.cpp#l2491
3. C2 compiler, e.g.: http://hg.openjdk.java.net/jdk8u/jdk8u/hotspot/file/c7a3e57fdf4a/src/cpu/x86/vm/x86_64.ad#l9859
4. If you happen to use Graal as the JIT compiler, Graal implements it intrinsically as well.

So while `StrictMath.log()` is always implemented by fdlibm in OpenJDK8u, `Math.log()` can be intrinsified into different implementations inside the OpenJDK HotSpot JVM that use platform-specific assembler, which can result in different results cross platform.

**Comment by srowen** [comment]
Yeah this is great info. I understand that java.lang.Math is allowed to return slightly different results from what some IEEE standard dictates but there is a most-correct answer in double precision, and these assembly code implementations in the x86 JVM seem to be clearly less correct for non-corner cases like log(3). That surprises me. There's a thread forwarded to dev@ with some additional info.

Anyway, I think it's a good move to make the actual log() function in SQL return values that are consistent platform and consistent with StrictMath, even if that doesn't mean we need to use StrictMath for every log call internally (for many usages in MLlib won't matter at all). And because there is basically no meaningful performance change.

**Comment by rednaxelafx** [comment]
I'd like to ask a post-hoc question on this PR: this one focuses on `log()` / `exp()` / `pow()` family of math functions. What about `sqrt()`? You can see that after this PR, there are still code like:
```scala
case class Acosh(child: Expression)
  extends UnaryMathExpression((x: Double) => StrictMath.log(x + math.sqrt(x * x - 1.0)), "ACOSH") {
```
left in mathExpressions.scala, which feels pretty inconsistent.

Are we planning to route `sqrt()` to `StrictMath` as well?

**Comment by srowen** [comment]
That could well be reasonable. I didn't do so simply because I wasn't aware of cases where the Oracle JDK returns different answers for sqrt(). At least, when running tests on ARM it evidently doesn't produce different answers for any of the sqrt-related tests, but that doesn't mean it doesn't exist.

**Pull request closed without merging**